include/range/v3/detail/config.hpp

```diff
@@ -18,19 +18,55 @@
 #include <iosfwd>
 
 #ifndef RANGES_ASSERT
+#if !defined(NDEBUG) && defined(__GNUC__) && !defined(__clang__) && __GNUC__ < 5
+#include <cstdio>
+#include <cstdlib>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        namespace detail
+        {
+            template<class = void>
+            [[noreturn]] void assert_failure(char const *file, int line, char const *msg)
+            {
+                std::fprintf(stderr, "%s(%d): %s\n", file, line, msg);
+                std::abort();
+            }
+        }
+    }
+}
+
+#define RANGES_ASSERT(...) ((__VA_ARGS__) ? void() : \
+    ::ranges::detail::assert_failure(__FILE__, __LINE__, "assertion failed: " #__VA_ARGS__))
+
+#else
 #include <cassert>
 #define RANGES_ASSERT assert
 #endif
+#endif
+
+#ifndef RANGES_ASSUME
+#if defined(__clang__) || defined(__GNUC__)
+#define RANGES_ASSUME(COND) ((COND) ? void() : __builtin_unreachable())
+#elif defined(_MSC_VER)
+#define RANGES_ASSUME(COND) ((void)(__assume(COND), (COND)))
+#else
+#define RANGES_ASSUME(COND) ((void)(COND))
+#endif
+#endif
 
 #ifndef RANGES_ENSURE_MSG
-#include <exception>
-#define RANGES_ENSURE_MSG(COND, MSG) \
-    ((COND) ? void() : (RANGES_ASSERT(!(true && MSG)), std::terminate()))
+#if defined(NDEBUG)
+#define RANGES_ENSURE_MSG(COND, MSG) RANGES_ASSUME(COND)
```

Comment:
This isn't right. An ENSURE is a check that is left in for Release builds. If the check fails, it should be a hard error. You've turned it into UB.